src/cascadia/TerminalApp/TabHeaderControl.cpp

```diff
@@ -0,0 +1,92 @@
+            _receivedKeyDown = true;
+        });
+
+        // NOTE: (Preview)KeyDown does not work here. If you use that, we'll
+        // remove the TextBox from the UI tree, then the following KeyUp
+        // will bubble to the NewTabButton, which we don't want to have
+        // happen.
+        HeaderRenamerTextBox().KeyUp([&](auto&&, Windows::UI::Xaml::Input::KeyRoutedEventArgs const& e) {
+            if (_receivedKeyDown)
+            {
+                if (e.OriginalKey() == Windows::System::VirtualKey::Enter)
+                {
+                    // User is done making changes, close the rename box
+                    _CloseRenameBox();
+                }
+                else if (e.OriginalKey() == Windows::System::VirtualKey::Escape)
+                {
+                    // User wants to discard the changes they made,
+                    // reset the rename box text to the old text and close the rename box
+                    HeaderRenamerTextBox().Text(Title());
+                    _CloseRenameBox();
+                }
+            }
+        });
+    }
+
+    // Method Description:
+    // - Updates the zoom icon indicating whether a pane is zoomed
+    // Arguments:
+    // - The desired visibility state of the zoom icon
+    void TabHeaderControl::SetZoomIcon(Windows::UI::Xaml::Visibility state)
+    {
+        HeaderZoomIcon().Visibility(state);
+    }
+
+    // Method Description:
+    // - Show the tab rename box for the user to rename the tab title
+    // - We automatically use the previous title as the initial text of the box
+    void TabHeaderControl::ConstructTabRenameBox()
+    {
+        _receivedKeyDown = false;
+
+        HeaderTextBlock().Visibility(Windows::UI::Xaml::Visibility::Collapsed);
+        HeaderRenamerTextBox().Visibility(Windows::UI::Xaml::Visibility::Visible);
+
+        HeaderRenamerTextBox().Text(Title());
+        HeaderRenamerTextBox().SelectAll();
+        HeaderRenamerTextBox().Focus(Windows::UI::Xaml::FocusState::Programmatic);
+    }
+
+    // Method Description:
+    // - Event handler for when the rename box loses focus
+    void TabHeaderControl::RenameBoxLostFocusHandler(Windows::Foundation::IInspectable const& /*sender*/,
+                                                     Windows::UI::Xaml::RoutedEventArgs const& /*e*/)
+    {
+        _CloseRenameBox();
+    }
+
+    // Method Description:
+    // - Hides the rename box and displays the title text block
+    // - Sends an event to the hosting tab to let them know we wish to change the title
+    //   to whatever is in the renamer box right now - the tab will process that event
+    //   and tell us to update our title
+    void TabHeaderControl::_CloseRenameBox()
+    {
+        HeaderRenamerTextBox().Visibility(Windows::UI::Xaml::Visibility::Collapsed);
+        HeaderTextBlock().Visibility(Windows::UI::Xaml::Visibility::Visible);
+
+        _HeaderTitleWantsToChangeHandlers(HeaderRenamerTextBox().Text());
```

Comment:
nit: maybe this needs to be one of those events where the sender just raises an event, and then the handler inspects the sender for the new value? IDK that seems like a lot of overhead for this pretty simple case though. @DHowett might have a stronger opinion.

Reply:
Nah, this one is fine.